### PR TITLE
Trigger a hard error on deprecation messages

### PR DESCRIPTION
**Pull request by wilzbach**
https://github.com/dlang/phobos/pull/5083#issuecomment-313140290

> Yeah but it doesn't make sense otherwise, does it? Seems antithetical to deprecate something for everyone else but not ourselves.
> Any reason not to enable `-de` on the test suite then?

CC @CyberShadow

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -5288,6 +5288,7 @@ if (isSomeString!S)
 }
 
 
+/*
 @safe pure @nogc unittest
```
Comment (CyberShadow) [inline std/string.d]: Wasn't it possible to mark the unittest as deprecated, which allows testing deprecated code? Or was I dreaming?

Comment (CyberShadow) [inline std/string.d]: Nope, that does seem to work!

**Comment by wilzbach** [comment]
> Wasn't it possible to mark the unittest as deprecated, which allows testing deprecated code? Or was I dreaming?
> Nope, that does seem to work!

Wow sweat - it works on my local machine :)

File: std/csv.d [hunk 2]
```diff
@@ -1099,7 +1099,6 @@ public:
     }
 }
 
-///
 @safe pure unittest
```
Comment (wilzbach) [inline std/csv.d]: `CsvReader` is `private`

File: std/uni.d [hunk 3]
```diff
@@ -2546,7 +2546,6 @@ private:
         return this;
     }
 
-    ///
     @safe unittest
     {
         assert(unicode.Cyrillic.intersect('-').byInterval.empty);
```
Comment (wilzbach) [inline std/uni.d]:
```
generated/linux/release/64/publictests/std_uni.d(132): Deprecation: std.uni.InversionList!(GcPolicy).InversionList.intersect(U)(U rhs) if (isCodepointSet!U) is not visible from module std_uni
```

File: std/uni.d [hunk 4]
```diff
@@ -4340,7 +4339,6 @@ if (sumOfIntegerTuple!sizes == 21)
     }
 }
 
-///
 @system pure unittest
 {
```
Comment (wilzbach) [inline std/uni.d]:
```
generated/linux/release/64/publictests/std_uni.d(192): Deprecation: std.uni.BitPacked(T, ulong sz) if (isIntegral!T || is(T : dchar)) is not visible from module std_uni
```

File: std/uni.d [hunk 5]
```diff
@@ -4644,7 +4642,6 @@ public struct MatcherConcept
         return this;
     }
 
-    ///
     @safe unittest
```
Comment (wilzbach) [inline std/uni.d]:
```
generated/linux/release/64/publictests/std_uni.d(242): Deprecation: std.uni.Utf8Matcher!().Impl!(1, 2, 3, 4).Impl.subMatcher(SizesToPick...)() is not visible from module std_uni
generated/linux/release/64/publictests/std_uni.d(243): Deprecation: std.uni.Utf8Matcher!().Impl!(1, 2, 3, 4).Impl.subMatcher(SizesToPick...)() is not visible from module std_uni
generated/linux/release/64/publictests/std_uni.d(244): Deprecation: std.uni.Utf8Matcher!().Impl!(1, 2, 3, 4).Impl.subMatcher(SizesToPick...)() is not visible from module std_uni
generated/linux/release/64/publictests/std_uni.d(245): Deprecation: std.uni.Utf8Matcher!().Impl!(1, 2, 3, 4).Impl.subMatcher(SizesToPick...)() is not visible from module std_uni
generated/linux/release/64/publictests/std_uni.d(250): Deprecation: std.uni.Utf16Matcher!().Impl!(1, 2).Impl.subMatcher(SizesToPick...)() is not visible from module std_uni
```

File: std/string.d [hunk 6]
```diff
@@ -5549,6 +5554,7 @@ S1 munch(S1, S2)(ref S1 s, S2 pattern) @safe pure @nogc
 }
 
 ///
+deprecated
 @safe pure @nogc unittest
```
Comment (CyberShadow) [inline std/string.d]: Yikes, why is this example publicly documented? We removed `munch`, so the example should go too

Comment (CyberShadow) [inline std/string.d]: Oops, never mind

**Comment by CyberShadow** [comment]
> I undocumented them for now to see whether it would pass all stages.

Yep, I think undocumenting them is the way to go for now. If the examples are not actually usable by users, then it's misleading to even show them. CI will prevent adding documented unittests that aren't usable outside the module, right?

**Comment by CyberShadow** [comment]
CC @DmitryOlshansky

**Comment by wilzbach** [comment]
> If the examples are not actually usable by users, then it's misleading to even show them. CI will prevent adding documented unittests that aren't usable outside the module, right?

Yup. Any idea on how we could tackle the deprecation warnings for Windows?
The problem is that `toUTF8/16` got only partially deprecated and these are legit uses:

```
std\windows\registry.d(48): Deprecation: function std.utf.toUTF8 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
std\windows\registry.d(48): Deprecation: function std.utf.toUTF16 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
std\file.d(2971): Deprecation: function std.utf.toUTF8 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
```

Maybe `by(W)Char.array`?

**Comment by CyberShadow** [comment]
Correct me if I'm wrong, but couldn't those uses simply be replaced with `std.conv.to!string` / `to!wstring`?

**Comment by wilzbach** [comment]
> Correct me if I'm wrong, but couldn't those uses simply be replaced with `std.conv.to!string` / `to!wstring`?

Let's see what the auto-tester says ;-)

**Comment by jmdavis** [comment]
> Yep, I think undocumenting them is the way to go for now. If the examples are not actually usable by users, then it's misleading to even show them.

Continuing to show the examples makes it easier for folks to figure out how the functions work so that they can more easily figure out how to fix their code so that it's not using the deprecated functions anymore - especially in cases where we're not providing a simple replacement (as is the case here). So, I think that it's better to keep the examples in the documentation if we can.

**Comment by CyberShadow** [comment]
> Continuing to show the examples makes it easier for folks to figure out how the functions work so that they can more easily figure out how to fix their code so that it's not using the deprecated functions anymore - especially in cases where we're not providing a simple replacement (as is the case here). So, I think that it's better to keep the examples in the documentation if we can.

In this particular case the problem was that the examples were using **private** symbols, so the examples were never actually usable. I think this coincides with enabling deprecation errors because previously it was possible to use private symbols in other modules under some circumstances.

To see for yourself, copy one of the unittests undocumented here to a new D file and try to compile them. E.g.:

```d
import std.uni;

void main()
{
	auto m = utfMatcher!char(unicode.Number);
	string square = "2²";
	// about sub-matchers
	assert(!m.subMatcher!(2,3,4).test(square)); // ASCII no covered
	assert(m.subMatcher!1.match(square)); // ASCII-only, works
	assert(!m.subMatcher!1.test(square)); // unicode '²'
	assert(m.subMatcher!(2,3,4).match(square));  //
	assert(square == "");
	wstring wsquare = "2²";
	auto m16 = utfMatcher!wchar(unicode.Number);
	// may keep ref, but the orignal (m16) must be kept alive
	auto bmp = m16.subMatcher!1;
	assert(bmp.match(wsquare)); // Okay, in basic multilingual plan
	assert(bmp.match(wsquare)); // And '²' too
}
```

This gives deprecation warnings because `subMatcher` is a private symbol, so there is something wrong either with the unittest or the declaration of `subMatcher`.

**Comment by jmdavis** [comment]
Ah, okay. That makes sense then.

**Comment by jmdavis** [comment]
> fgsfds

Did github not let you provide an empty comment when dismissing your review?

**Comment by CyberShadow** [comment]
Exactomundo.

**Comment by CyberShadow** [comment]
Some remaining ones (from the autotester):

```
std\file.d(2647): Deprecation: function std.utf.toUTF8 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
std\file.d(3433): Deprecation: Comparison between different enumeration types `Flag` and `Flag`; If this behavior is intended consider using `std.conv.asOriginalType`
std\file.d(4289): Deprecation: function std.utf.toUTF8 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
std\datetime\timezone.d(2994): Deprecation: function std.utf.toUTF16 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
std\windows\charset.d(117): Deprecation: function std.utf.toUTF8 is deprecated - To be removed November 2017. Please use std.utf.encode instead.
```

File: std/windows/charset.d [hunk 7]
```diff
@@ -53,7 +53,6 @@ version (Windows):
 private import core.sys.windows.windows;
 private import std.conv;
 private import std.string;
-private import std.utf;
 private import std.windows.syserror;
```
Comment (wilzbach) [inline std/windows/charset.d]: Btw what's the reason for `private import`s? Aren't they private by default?

Comment (CyberShadow) [inline std/windows/charset.d]: I think very very old versions of D (well into 0.x) had imports public by default, like C header files.

Comment (wilzbach) [inline std/windows/charset.d]: Wow -> https://github.com/dlang/phobos/pull/5553

**Comment by wilzbach** [comment]
Yeah it's finally passing on Windows:

image

(I rebased my three attempts to get Windows working into one commit)

**Comment by CyberShadow** [comment]
LGTM. @DmitryOlshansky Please have a look at the std.uni unit tests when you find the time.

**Comment by BBasile** [comment]
Pardon me but why did you ( @wilzbach @CyberShadow @jmdavis )  make and/or accept this completely stupid thing ? Now when the goal of a DMD pull request is to deprecate something everything goes red and the auto tester cant be used anymore to fix the PR (e.g platform specific failure missed locally) because the tests don't run anymore.

see the tests here https://github.com/dlang/dmd/pull/6925

:clap:

**Comment by wilzbach** [comment]
> Now when the goal of a DMD pull request is to deprecate something everything goes red and the auto tester cant be used anymore to fix the PR (e.g platform specific failure missed locally) because the tests don't run anymore.

That's exactly why we made this PRs in case you haven't noticed as there have been continuous incidents (aka __regressions__) of deprecations within Phobos showing up in the user build due to Posix being the most popular developer platform.

See also: https://github.com/dlang/phobos/pull/5083#issuecomment-282540482

**Comment by BBasile** [comment]
But it can't work. Phobos and DMD are two different repositories !

**Comment by mathias-lang-sociomantic** [comment]
I wholeheartedly agree with @BBasile here, but for different reason. If you have code which have a certain behavior, and you want to deprecate one symbol, but keep the same behavior, there are places where you have no choices but to use a deprecated symbol.

One example which happened recently is [here](https://github.com/sociomantic-tsunami/ocean/pull/173). I ended up using a hack with `extern(C)`, but really we should have a way to allow library users to use deprecated feature when they really need to.

**Comment by CyberShadow** [comment]
> I wholeheartedly agree with @BBasile here, but for different reason. If you have code which have a certain behavior, and you want to deprecate one symbol, but keep the same behavior, there are places where you have no choices but to use a deprecated symbol.

I suggest to wait until we run into a situation that warrants reconsidering this before reverting it.

> One example which happened recently is here. I ended up using a hack with extern(C), but really we should have a way to allow library users to use deprecated feature when they really need to.

Why not:

1. Move the implementation of the public function that is to be deprecated to a private / package `xxxImpl` function
2. Add a deprecated wrapper or alias which forwards to the `xxxImpl` function
3. Change all internal references of `xxx` to `xxxImpl` ?

**Comment by mathias-lang-sociomantic** [comment]
> Why not: [...]

It will work, but you are basically duplicating all your public symbols, as you need to do it for everything you are deprecating. It feels very wrong that you have to do such a large patchset for a simple deprecation. But yes, it's possible if you are willing to go through that pain.

**Comment by CyberShadow** [comment]
I don't know; it seems more wrong to me to be hypocritical about deprecations.

If you absolutely need to use public symbols from inside your implementations so much that deprecating such public symbols becomes an issue, it may indicate that there is a problem with the separation of interface and implementation in your code base. Futhermore, projects that wish to use the `-de` switch will not be able to directly use (e.g. via rdmd) libraries which rely on this behavior.

Anyway, the requirements and internal structure of Ocean and Phobos are undoubtedly very different, so I suggest to wait and see if/when we run into an actual problem.

**Comment by ZombineDev** [comment]
@mathias-lang-sociomantic I don't understand your reaction. What's the problem with making deprecations errors inside Phobos? How does this effect anyone else? Of course, if we reach a place where this is a problem, this could easily be disabled, but for the time being, I believe that the more rigorous testing is well worth it. 

@BBasile what's the problem with fixing the errors caused by deprecations first and then merging PRs that actually introduce those deprecations? Either way, the hardest thing is getting the proposal to deprecate something accepted rather than fixing the stuff themselves, IMO (though your specific case is a bit more involved because people were casting string literals to static arrays).

**Comment by CyberShadow** [comment]
Case in point:
https://issues.dlang.org/show_bug.cgi?id=17621

**Pull request merged**